### ISSUE TITLE
[Enhancement] Button/Label Parity

### ISSUE THREAD

**Issue by kingces95** [xamarin/Xamarin.Forms#1668]
## Rationale

A Button containing text is basically a fancy Label; Options for formatting text on Label make sense to expose on Button. Specifically, justification and line wrapping options are available on Label but not on Button.

see https://forums.xamarin.com/discussion/36779/multi-line-text-on-buttons-how-to-get-consistent-across-platforms

## Implementation

Adopt same API on Label for text justification and line wrapping. At least expose `HorizontalTextAlignmentProperty`, `VerticalTextAlignmentProperty`, and `LineBreakModeProperty`.

## Expected Result

User is able to justify and control button text using label API. 

## Implications for CSS

None.

## Backward Compatibility

As discussed in the thread, the different platforms have different defaults. We may have to introduce new enum values like `Legacy` or `Undefined` or `Default`. 

## Difficulty : Moderate

Anything with potential backwards compatibility issues is moderately difficult. Also, the text matrix for this work would be large and clip detection is non-trivial.

**Comment by hartez** [comment]
@kingces95 @samhouts This appears to mostly be a duplicate of https://github.com/xamarin/Xamarin.Forms/issues/1659. The only difference I'm seeing is that this issue calls for `LineBreakMode` support. Should this issue be updated to include only `LineBreakMode`?

**Comment by ice-j** [comment]
@davidortinau I would like to take this issue too, since I'm working on #1659

**Comment by davidortinau** [comment]
@ice-j any progress on this? Any blockers we can help remove?

**Comment by ice-j** [comment]
@davidortinau https://github.com/XamarinFormsCommunity/Xamarin.Forms/pull/2#discussion_r175752527 

The way I have implemented it, it also aligns the image (when set) on the button. I'm waiting on any ideas if you have to approach this differently.

**Comment by samhouts** [comment]
@ice-j Are we ready to move this pull request to the main repository?

**Comment by samhouts** [comment]
@ice-j Monthly ping! Would you like to submit the pull request here? Thanks!

**Comment by ice-j** [comment]
Oh, sorry, there was an unanswered question on the community PR, that's why I didn't move it here, but I will do this weekend and maybe I can continue work here if anyone else doesn't want to take the ticket.

**Comment by StephaneDelcroix** [comment]
@ice-j ping

**Comment by groege** [comment]
@ice-j @StephaneDelcroix 
Ping - It's been a while, is there any news?

**Comment by Mikilll94** [comment]
Ping again...

**Comment by Happypig375** [comment]
I can't believe that Xamarin has little to no intention of providing such a basic feature of all UI frameworks.

**Comment by pictos** [comment]
@Mikilll94 and @Happypig375 I've a PR related to this issue. You can see it [here](https://github.com/xamarin/Xamarin.Forms/pull/7810).

You can help the team doing a code review ;)

**Comment by EmilAlipiev** [comment]
any news on this?

**Comment by Happypig375** [comment]
@pictos `HorizontalTextAlignment` and `VerticalTextAlignment` are still missing.

**Comment by pictos** [comment]
@Happypig375 I know. For now that PR is solving this [issue](https://github.com/xamarin/Xamarin.Forms/issues/3106), which is related to this one. After that PR gets merged, I can work in `HorizontalTextAlignment` and `VerticalTextAlignment`

**Comment by Happypig375** [comment]
@pictos I don't think it being removed from the "In Progress" column of vNext is a sign of merging.

**Comment by titoleru** [comment]
Ping - what is the current status of this?

**Comment by nonubitta** [comment]
Its sad, such a basic feature already took more than two years and still no clue. 👎 
:(

**Comment by titoleru** [comment]
come on, we missing the HorizontalTextAlignment & VerticalTextAlignment properties in the Button. Please any updates on this?

**Comment by ElVinche** [comment]
Hello, any news about this issue? it's really annoying in some case

**Comment by d617617** [comment]
If you use button in grid,the btton's text  will be in left. it is bad!we need this change!

**Comment by Tommigun1980** [comment]
I think it's quite bad that one of the most frequently used UI elements, a button, does not allow to align its text. These kinds of issues should be absolute top priority for the Xamarin team in my opinion, as it is almost impossible to do anything with Xamarin.Forms without re-creating the entire set of UI widgets. I can not understand how this issue has not even been started yet.

Is there any way the team could fix this as I am getting very tired of implementing primitive UI widgets all the time? I thought this was the exact problem Xamarin.Forms was supposed to solve.

Thank you for considering.

**Comment by Tommigun1980** [comment]
How are the sprints planned for the Xamarin.Forms team? Are people working on their pet projects, or is there a real vetting process? Please focus on the built-in components to have at least some kind of feature parity with the platforms.

**Comment by Tommigun1980** [comment]
> Its sad, such a basic feature already took more than two years and still no clue. 👎
> :(

I think that the Xamarin.Forms team must be freestyling with things they want to work on, like the "Shell". It's a good example of something they put a lot of resources in which has no real benefit, as it is usable only if you make an application with exactly the components they felt like implementing. While at the same time the primitive UI controls are almost unusable. So I don't think this will get implemented as nobody feels like working on it, even though their product is quite unusable without a base set of features.

**Comment by Happypig375** [comment]
If there is one thing Xamarin should focus on, it is the [1606 bugs that are still open!](https://github.com/xamarin/Xamarin.Forms/issues?q=is%3Aopen+is%3Aissue+label%3A%22t%2Fbug+%3Abug%3A%22)

**Comment by Happypig375** [comment]
The fact that Xamarin.Forms has roughly the same amount of bugs as [open issues of mono](https://github.com/mono/mono/issues) really speaks for itself.

**Comment by Tommigun1980** [comment]
> If there is one thing Xamarin should focus on, it is the [1606 bugs that are still open!](https://github.com/xamarin/Xamarin.Forms/issues?q=is%3Aopen+is%3Aissue+label%3A%22t%2Fbug+%3Abug%3A%22)

Bugs are #1 for sure, that goes without saying. I was referring to development of new features. In that regard I really think they should focus on making the CORE work well.

**Comment by d617617** [comment]
> > 如果Xamarin应该着重解决一件事，那就是[1606个bug仍然存在！](https://github.com/xamarin/Xamarin.Forms/issues?q=is%3Aopen+is%3Aissue+label%3A%22t%2Fbug+%3Abug%3A%22)
> 
> 错误肯定是[第](https://github.com/xamarin/Xamarin.Forms/pull/1)一位的，这毋庸置疑。我指的是开发新功能。在这方面，我真的认为他们应该集中精力使CORE运作良好。
i think it is not possible,you can look the new paltform which name maui.

**Comment by Hantick** [comment]
Still waiting for aligment

**Comment by AlaJaber97** [comment]
any progress on this?

**Comment by BlueRaja** [comment]
The documentation for [`Button.ContentLayout`](https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.button.contentlayout?view=xamarin-forms) says

> Gets or sets an object that controls the position of the button image **and the spacing between the button's image and the button's text.**

This is actually not true since **the text is always centered**, so the spacing is actually between the button and the left-edge of the box the text is being centered in (leading to every button with an image having off-center text).

This has been known for at least 5 years. It's insane that this still hasn't been fixed...

**Comment by jfversluis** [comment]
We won't be investing in this for Xamarin.Forms anymore. I think steps have already been made with this in .NET MAUI. If this is still important to you, make sure to voice any concerns on the .NET MAUI repo. Thanks!

**Comment by HaufenKlaus** [comment]
> We won't be investing in this for Xamarin.Forms anymore. I think steps have already been made with this in .NET MAUI. If this is still important to you, make sure to voice any concerns on the .NET MAUI repo. Thanks!

and you want to make us believe that oyu actualyl do that after at least 3 years of kicking this ez implementation down the road?

> The documentation for [`Button.ContentLayout`](https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.button.contentlayout?view=xamarin-forms) says
> 
> > Gets or sets an object that controls the position of the button image **and the spacing between the button's image and the button's text.**
> 
> This is actually not true since **the text is always centered**, so the spacing is actually between the button and the left-edge of the box the text is being centered in (leading to every button with an image having off-center text).
> 
> This has been known for at least 5 years. It's insane that this still hasn't been fixed...

i also got the feeling that they don't want to ruin the business model of third party component providers. It's the same with unity: have an abysmally outdated product, let a third party swoop in, make customer pay. I can already smell the nepotism

**Comment by HaufenKlaus** [comment]
for anyone still running into this issue: here's a hacky and Xamarin.Forms worthy workaround:

Use a Label like this. You can style it like a button and don't need to write extra code for Android and iOS, bc we shouldn't have a DX from the 90s...
```
<Label
    <Label.GestureRecognizers>
      <TapGestureRecognizer Command="{Binding YourCommand}" />
    </Label.GestureRecognizers>
</Label>
```
